### ISSUE TITLE
Request ID token option for google_sign_in plugin

### ISSUE THREAD

**Issue by deckerst** [flutter/flutter#20903]
With google_sign_in v3.0.4, there is no option to request an ID token.

On iOS, the ID token is always returned, but on Android it is null unless the builder option is provided.

Here is the [documentation](https://developers.google.com/android/reference/com/google/android/gms/auth/api/signin/GoogleSignInOptions.Builder.html#requestIdToken(java.lang.String)) of the missing option method.

**Comment by deckerst** [comment]
It seems that as of google_sign_in v3.2.1, the ID token is returned automatically on Android too, just like on iOS, so the builder option is not necessary.

Shall I close this ticket, or some work is planned to provide that option anyway to match the API?

**Comment by thehitechpanky** [comment]
If you are using the app without firebase it always return a null id_token

**Comment by kroikie** [comment]
@deckerst

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/693. Any further collaboration will be done there.

**Comment by lucavenir** [comment]
I didn't expect this issue to have 5 years!

How do people work around this issue reliably? I _have to_ send this token to my server, I am confused as to why this isn't possible with Android.

I've tried everything I've found on the web (and on this repo), but nothing seems to work if you're not using Firebase, which is clearly a business decision.

Any update on when this issue will be properly solved / documented?

**Comment by shaynec25** [comment]
Had the same issue, didn't use Firebase.
@lucavenir 's solution helps! thanks a lot!

**Comment by aidbull** [comment]
@lucavenir thank you very much, I've lost so many brain cells stressing out over this issue. 
God bless you.

**Comment by PascalMalz** [comment]
OmG after so many hours of tests and research this approach works.

**Comment by PlugFox** [comment]
@lucavenir thanks you **SO MUCH**

1) Create a new Credential at Google Console for the Web OAuth2 project
And Android credential with SHA1 and project ID, too!!!
You need both of these!

2) Add android dependencies `android\app\build.gradle` like this:

```gradle
...
dependencies {
  implementation 'com.google.android.gms:play-services-auth:20.7.0'
}
```

3) Implement authentication with dart:
You should use a **Web credential-related Client ID**, not an Android one!
![image](https://github.com/flutter/flutter/assets/7805176/06e25ee1-343c-4a95-81b2-7a080919ee79)

**Comment by superern** [comment]
> I didn't expect this issue to be 5 years old: any update on when this issue will be properly solved and, most importantly, well documented even for non-firebase users?
> 
> **EDIT.** Ok, I hadn't the most fun days but I managed to at least reach a reliable workaround (😭). In case anyone is still having trouble on this (i.e. using `GoogleSignIn` _without_ Firebase), to solve this Android issue I've been greatly helped by this [documentation](https://developers.google.com/identity/one-tap/android/get-started), combined with [this comment](https://github.com/flutter/flutter/issues/12140#issuecomment-651513048).
> 
> Shortly, in my case, I had to:
> 
> 1. Create a empty, non-configured Web OAuth Token Credential in the Google Cloud console, (do this, **even if you don't actually use a web client**)
> 2. Paste its `clientId` inside a properly formatted `android/app/src/[flavor]/res/values/strings.xml` (again, see the linked comment above)
> 3. Add `implementation 'com.google.android.gms:play-services-auth:20.4.1'` in in `android/app/build.gradle`, under the `dependencies` property (see the linked docs above if in doubt)
> 
> Hope this helps anyone's frustration as much as it helped me

This really works!

Create Credentials ( Oauth Client Id )
[Google Cloud Console](https://console.cloud.google.com/apis/credentials)
<img width="458" alt="image" src="https://github.com/flutter/flutter/assets/39351808/5f639170-4d2b-4142-b172-de403c5ab2eb">

Choose WebApplication
https://console.cloud.google.com/apis/credentials

Dont do anything, just click Save.

Open the newly created oauth and copy the clientId

then update this .xml file (`android/app/src/[flavor]/res/values/strings.xml`)
https://github.com/flutter/flutter/issues/12140#issuecomment-651513048

**Comment by Douoo** [comment]
Anyone facing this issue on the web version. I tried everything from setting my OAuth key on google cloud console to my flutter app but nothing seems to work. The idToken is null. I'm wondering if this issue is from flutter web itself. Please somebody let me know if there is a fix to it

**Comment by specOper99** [comment]
@Douoo As you can see in the documentation of `google_sign_in_web` there is no way to obtain the `idToken` and you should depend on the `accessToken` in order to check your user state on the server

<img width="806" alt="Screenshot 2023-12-17 at 10 18 35 AM" src="https://github.com/flutter/flutter/assets/59845618/42a68772-961f-4c4c-8cda-a1ca873c1422">

**Comment by sbatezat** [comment]
> @Douoo As you can see in the documentation of `google_sign_in_web` there is no way to obtain the `idToken` and you should depend on the `accessToken` in order to check your user state on the server
> 
> <img alt="Screenshot 2023-12-17 at 10 18 35 AM" width="806" src="https://private-user-images.githubusercontent.com/59845618/291033730-42a68772-961f-4c4c-8cda-a1ca873c1422.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3ODQ0NDMsIm5iZiI6MTcwMzc4NDE0MywicGF0aCI6Ii81OTg0NTYxOC8yOTEwMzM3MzAtNDJhNjg3NzItOTYxZi00YzRjLThjZGEtYTFjYTg3M2MxNDIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDE3MjIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5ZWQ0ZDVlMTY4OTFjM2IyMThiNzY3OWM0ZTZhYTJhYWQ4YzkzNmVhMjZhYzI2NmY1MzRiYzdmMzdjNmU4NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EuDNVA_fIkVuWkWmoQuO5349QqU89dxSDpiMOWW35XQ">

It's wrong. You can obtain an Id Token if you use the renderButton widget, as documented :

<img width="777" alt="image" src="https://github.com/flutter/flutter/assets/10560159/5c14ee07-b7b6-4db2-aa89-941e72e3cc4e">

I'm currently using "google-auth-library" server side (nodeJS) to verify these id token :
```

const ticket = await this.googleAuthClient.verifyIdToken({
                idToken: token,
                audience: [
                    this.GOOGLE_ANDROID_CLIENT_ID,
                    this.GOOGLE_IOS_CLIENT_ID,
                    this.GOOGLE_WEB_CLIENT_ID,
                ],
            });
```

I'm curious @specOper99 what's the equivalent with access token ? i.e : do not just verify it's a valid access token, but also is has been delivered by my app?

**Comment by infitry** [comment]
Enter the client_id whose client_type of oauth_client in google-services.json is 3 
`<string name="default_web_client_id">{here}</string>`

Please replace it with {here}.

Also, add the corresponding client_id to the audience of GoogleIdTokenVerifier on the server side
If there is no oauth_client with client_type 3, just add web app in firebase.

**Comment by nayanAubie** [comment]
https://github.com/flutter/flutter/issues/20903#issuecomment-1978715265
This solution is working, I've tested with `4.3.15` and `4.4.0` with the same code base. with `4.4.0`, I am getting `idToken = null` and in `4.3.15`, I am getting the value in `idToken`.

**Comment by nayanbabariya** [comment]
https://github.com/flutter/flutter/issues/20903#issuecomment-1978715265 This solution is working, I've tested with `4.3.15` and `4.4.0` with the same code base. with `4.4.0`, I am getting `idToken = null` and in `4.3.15`, I am getting the value in `idToken`.

**Comment by akshaytaru2007** [comment]
@nayanbabariya I agree. Same thing happened for me. Even if we use `com.google.gms:google-services:4.4.2` then also same thing happening.

**Comment by s4rath07** [comment]
Is there any solution for this problem which is happening in the flutter web. For the android it is working. But for web it is giving null.

**Comment by stuartmorgan** [comment]
Closing as a duplicate of https://github.com/flutter/flutter/issues/36673, as this is a documentation issue rather than an API issue; we have a mechanism to return the ID token, it's just not sufficiently clear how to configure the requests when not using the Firebase path.
